### PR TITLE
Adjust `-Ctarget-cpu=native` handling in cg_llvm

### PR DESCRIPTION
**Pull request by nagisa**
When cg_llvm encounters the `-Ctarget-cpu=native` it computes an
explciit set of features that applies to the target in order to
correctly compile code for the host CPU (because e.g. `skylake` alone is
not sufficient to tell if some of the instructions are available or
not).

However there were a couple of issues with how we did this. Firstly, the
order in which features were overriden wasn't quite right – conceptually
you'd expect `-Ctarget-cpu=native` option to override the features that
are implicitly set by the target definition. However due to how other
`-Ctarget-cpu` values are handled we must adopt the following order
of priority:

* Features from -Ctarget-cpu=*; are overriden by
* Features implied by --target; are overriden by
* Features from -Ctarget-feature; are overriden by
* function specific features.

Another problem was in that the function level `target-features`
attribute would overwrite the entire set of the globally enabled
features, rather than just the features the
`#[target_feature(enable/disable)]` specified. With something like
`-Ctarget-cpu=native` we'd end up in a situation wherein a function
without `#[target_feature(enable)]` annotation would have a broader
set of features compared to a function with one such attribute. This
turned out to be a cause of heavy run-time regressions in some code
using these function-level attributes in conjunction with
`-Ctarget-cpu=native`, for example.

With this PR rustc is more careful about specifying the entire set of
features for functions that use `#[target_feature(enable/disable)]` or
`#[instruction_set]` attributes.

Sadly testing the original reproducer for this behaviour is quite
impossible – we cannot rely on `-Ctarget-cpu=native` to be anything in
particular on developer or CI machines.

cc https://github.com/rust-lang/rust/issues/83027 @BurntSushi

### CODE REVIEW

**Comment by BurntSushi** [comment]
Thank you @nagisa!!!

File: compiler/rustc_codegen_llvm/src/llvm_util.rs [hunk 1]
```diff
@@ -218,13 +218,34 @@ pub fn target_cpu(sess: &Session) -> &str {
     handle_native(name)
 }
 
-pub fn handle_native_features(sess: &Session) -> Vec<String> {
+/// The list of LLVM features computed from CLI flags (`-Ctarget-cpu`, `-Ctarget-feature`,
+/// `--target` and similar).
+// FIXME(nagisa): Cache the output of this somehow? Maybe make this a query? We're calling this
+// for every function that has `#[target_feature]` on it. The global features won't change between
```
Comment (lqd) [inline compiler/rustc_codegen_llvm/src/llvm_util.rs:224]: was this function called as often before, by the way ? we should probably run the perfbot on this PR, as there are thousands of vendor and SIMD intrinsics with `#[target_feature]` in `stdarch`, just in case this has become more expensive.

File: compiler/rustc_codegen_llvm/src/attributes.rs [hunk 2]
```diff
@@ -301,20 +289,22 @@ pub fn from_fn_attrs(cx: &CodegenCx<'ll, 'tcx>, llfn: &'ll Value, instance: ty::
     // The target doesn't care; the subtarget reads our attribute.
     apply_tune_cpu_attr(cx, llfn);
 
-    let features = llvm_target_features(cx.tcx.sess)
```
Comment (nagisa) [inline compiler/rustc_codegen_llvm/src/attributes.rs:304]: @lqd in answer to your question, we would call the `llvm_target_features` function for every single function call in the past, but it was somewhat(?) cheaper in that it would be able to produce an iterator of `&str`s, rather than a `Vec<String>` as `llvm_global_features` is forced to now.

However as you see in this area, we'd go on to immediately convert each of those `&str`s to a `String` anyway. As thus I wouldn't expect much difference in performance.

Comment (lqd) [reply compiler/rustc_codegen_llvm/src/attributes.rs:304]: makes sense, thanks for the answer !

File: compiler/rustc_codegen_llvm/src/llvm_util.rs [hunk 3]
```diff
@@ -218,13 +218,34 @@ pub fn target_cpu(sess: &Session) -> &str {
...
+// the functions; only crates, maybe…
+pub fn llvm_global_features(sess: &Session) -> Vec<String> {
+    // FIXME: this should definitely be available more centrally and to other codegen backends.
+    /// These features control behaviour of rustc rather than llvm.
+    const RUSTC_SPECIFIC_FEATURES: &[&str] = &["crt-static"];
+
+    // Features that come earlier are overriden by conflicting features later in the string.
+    // Typically we'll want more ecplicit settings to override the implicit ones, so:
```
Comment (petrochenkov) [inline compiler/rustc_codegen_llvm/src/llvm_util.rs:232]:
```suggestion
    // Typically we'll want more explicit settings to override the implicit ones, so:
```

File: compiler/rustc_codegen_llvm/src/llvm_util.rs [hunk 4]
```diff
@@ -218,13 +218,34 @@ pub fn target_cpu(sess: &Session) -> &str {
...
+    //
+    // * Features implied by --target; are overriden by
+    // * Features from -Ctarget-cpu=native; are overriden by
```
Comment (petrochenkov) [inline compiler/rustc_codegen_llvm/src/llvm_util.rs:235]: How does `-Ctarget-cpu=something-specific` fit into this picture?

Comment (nagisa) [reply compiler/rustc_codegen_llvm/src/llvm_util.rs:235]: Hm, that's a good question. Today we only pass a computed list of `target-feature`s for `-Ctarget-cpu=native` only.

For other `-Ctarget-cpu`s the `target-feature` list we pass to LLVM will only contain values from `--target` and `-Ctarget-feature` flag. The features present in the string we pass to LLVM will override those that are implied from `-Ctarget-cpu` (regardless of whether it is set implicitly through target or explicitly via CLI).

In short, there ends up being a difference in feature priority: `-Ctarget-cpu=!native` is overriden by `target-features` implied by `--target` is overriden by `-Ctarget-cpu=native` is overriden by `-Ctarget-features`.

I think we still want `-Ctarget-cpu` to override the features implied by `--target`, but that sounds difficult to handle in all cases (including the one where no `-Ctarget-cpu` flag is specified at all...)

I'll sleep on this.

Comment (nagisa) [reply compiler/rustc_codegen_llvm/src/llvm_util.rs:235]: I'm going to change it back to the previous ordering (`-Ctarget-cpu` -> `--target` -> `-Ctarget-features`) just to facilitate landing this sooner and figure out something about the ordering otherwise later.

**Comment by petrochenkov** [comment]
PR description needs to be updated.
r=me after that.

**Comment by nagisa** [comment]
@bors r=petrochenkov

(also updated the description in the commit message)

**Comment by nagisa** [comment]
@bors r=petrochenkov

(was missing a `--target` in one of the tests)

**Comment by camelid** [comment]
Does this fix #83027?

**Comment by camelid** [comment]
If so, this PR should probably be backported to beta so the bug isn't in 1.51.

**Comment by nagisa** [comment]
I didn't test the original reproducer, remains for the others to verify.

**Comment by pnkfelix** [comment]
did a quick beta nomination and attempted to discuss at T-compiler meeting today, but the meeting was over time at that point so we didn't get much discussion.

collectively I'd say we are hesitant at best to beta-backport this change. As nagisa put it in the meeting:

>  this problem only affects a small number of people using a very specific flag combination which isn't a default and has been known to break in other ways in the past (miscompile)

I'm leaving the beta-nomination label on it at the moment while we wait for @BurntSushi to respond about whether PR #83084 fixes #83027. But even if it *does* fix issue #83027, it seems likely that we'll still decline to backport; its a big patch and somewhat risky, which is a tough tradeoff given context established by nagisa above.

**Comment by apiraino** [comment]
Beta backport declined ([zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202021-03-18.20.2354818/near/230878420))

@rustbot label -beta-nominated

**Comment by BurntSushi** [comment]
@pnkfelix I agree with not back-porting. If this didn't require `target-cpu=native` to be triggered, that'd be a different story. But `target-cpu=native` definitely makes this lower impact!

**Comment by pnkfelix** [comment]
(hmm can someone tell me why rustbot removed the beta-nomination up above? Was that because of #83027 getting closed somehow...?)

**Comment by apiraino** [comment]
> (hmm can someone tell me why rustbot removed the beta-nomination up above? Was that because of #83027 getting closed somehow...?)

@pnkfelix heh ... github web UI sometimes fails to update comments. I had wrote a comment to `@rustbot label -beta-nominated` 7 minutes after you [wrote yours](https://github.com/rust-lang/rust/pull/83084#issuecomment-802011888) but my page didn't refresh. So I have removed my comment and [manually reinstated](https://github.com/rust-lang/rust/pull/83084#event-) the ` beta-nominated` label. Which you [then removed](https://github.com/rust-lang/rust/pull/83084#event-) again :smile:
